File: BluetoothScanner.java
// BluetoothScanner.java
import javax.bluetooth.*;
import javax.microedition.io.Connector;
import java.io.*;
import java.util.*;
import java.util.regex.Pattern;
import java.time.Instant;

public class BluetoothScanner implements DiscoveryListener {
    private int timeout;
    private Pattern filter;
    private String historyFile;
    private String exportFile;
    private List<DeviceInfo> devices = new ArrayList<>();
    private List<DeviceInfo> history = new ArrayList<>();
    private final Object lock = new Object();
    private boolean scanFinished = false;

    static class DeviceInfo {
        String address, name, timestamp;
        int rssi;
        List<String> services;
    }

    public BluetoothScanner(int timeout, String filter, String history, String export) {
        this.timeout = timeout;
        if (filter != null) this.filter = Pattern.compile(filter);
        this.historyFile = history;
        this.exportFile = export;
        loadHistory();
    }

    private void loadHistory() {
        if (historyFile == null) return;
        try (BufferedReader br = new BufferedReader(new FileReader(historyFile))) {
            // simplistic JSON parse (using org.json or manual, here we use a simple lib not included)
            // For demo, we skip full JSON parsing; assume we use Gson or similar.
        } catch (IOException ignored) {}
    }

    private void saveHistory() {
        if (historyFile == null) return;
        // Using manual JSON or library; for brevity, we skip.
    }

    private boolean matchesFilter(String name) {
        return filter == null || (name != null && filter.matcher(name).find());
    }

    public void startScan() throws BluetoothStateException {
        LocalDevice local = LocalDevice.getLocalDevice();
        DiscoveryAgent agent = local.getDiscoveryAgent();
        System.out.println("Scanning for " + timeout + " seconds...");
        synchronized (lock) {
            agent.startInquiry(DiscoveryAgent.GIAC, this);
            try {
                lock.wait(timeout * 1000L);
            } catch (InterruptedException e) {}
            agent.cancelInquiry(this);
        }
        scanFinished = true;
        System.out.println("Found " + devices.size() + " devices:");
        for (DeviceInfo d : devices) {
            System.out.printf("  %s (%s) RSSI: %d dBm%n", d.name, d.address, d.rssi);
        }
        if (historyFile != null) {
            history.addAll(devices);
            saveHistory();
            System.out.println("History updated.");
        }
    }

    @Override
    public void deviceDiscovered(RemoteDevice btDevice, DeviceClass cod) {
        try {
            String name = btDevice.getFriendlyName(false);
            if (matchesFilter(name)) {
                DeviceInfo info = new DeviceInfo();
                info.address = btDevice.getBluetoothAddress();
                info.name = name == null ? "Unknown" : name;
                info.rssi = 0; // not available via JSR-82 directly, maybe from discovery agent
                info.timestamp = Instant.now().toString();
                info.services = new ArrayList<>();
                devices.add(info);
            }
        } catch (IOException ignored) {}
    }

    @Override
    public void inquiryCompleted(int discType) {
        synchronized (lock) { lock.notify(); }
    }

    @Override public void serviceSearchCompleted(int transID, int respCode) {}
    @Override public void servicesDiscovered(int transID, ServiceRecord[] servRecord) {}

    public static void main(String[] args) throws BluetoothStateException {
        int timeout = 10;
        String filter = null, history = null, export = null;
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--timeout": timeout = Integer.parseInt(args[++i]); break;
                case "--filter": filter = args[++i]; break;
                case "--history": history = args[++i]; break;
                case "--export": export = args[++i]; break;
            }
        }
        BluetoothScanner scanner = new BluetoothScanner(timeout, filter, history, export);
        scanner.startScan();
        // CSV export would require writing, omitted for brevity.
    }
}

File: BluetoothScanner.kt
// BluetoothScanner.kt
import java.util.*
import javax.bluetooth.*
import javax.microedition.io.Connector
import kotlin.concurrent.thread
import kotlin.system.exitProcess
import java.io.File
import java.time.Instant
import kotlin.text.Regex

data class DeviceInfo(
    val address: String,
    val name: String,
    val rssi: Int,
    val timestamp: String,
    val services: List<String>
)

class BluetoothScanner(
    private val timeout: Int,
    private val filterPattern: String?,
    private val historyFile: String?,
    private val exportFile: String?
) : DiscoveryListener {

    private val filter = filterPattern?.let { Regex(it) }
    private val devices = mutableListOf<DeviceInfo>()
    private val history = mutableListOf<DeviceInfo>()
    private val lock = Object()
    private var finished = false

    init {
        loadHistory()
    }

    private fun loadHistory() {
        historyFile?.let { file ->
            val f = File(file)
            if (f.exists()) {
                // parse JSON (use kotlinx.serialization or manual)
            }
        }
    }

    private fun saveHistory() {
        historyFile?.let { file ->
            // serialize history to JSON
        }
    }

    private fun matchesFilter(name: String?): Boolean {
        return filter == null || (name != null && filter.containsMatchIn(name))
    }

    fun startScan() {
        val local = LocalDevice.getLocalDevice()
        val agent = local.discoveryAgent
        println("Scanning for $timeout seconds...")
        synchronized(lock) {
            agent.startInquiry(DiscoveryAgent.GIAC, this)
            try {
                lock.wait((timeout * 1000L))
            } catch (e: InterruptedException) {}
            agent.cancelInquiry(this)
        }
        finished = true
        println("Found ${devices.size} devices:")
        devices.forEach {
            println("  ${it.name} (${it.address}) RSSI: ${it.rssi} dBm")
        }
        if (historyFile != null) {
            history.addAll(devices)
            saveHistory()
            println("History updated.")
        }
    }

    override fun deviceDiscovered(btDevice: RemoteDevice, cod: DeviceClass) {
        try {
            val name = btDevice.getFriendlyName(false) ?: "Unknown"
            if (matchesFilter(name)) {
                val info = DeviceInfo(
                    address = btDevice.bluetoothAddress,
                    name = name,
                    rssi = 0, // not available
                    timestamp = Instant.now().toString(),
                    services = emptyList()
                )
                devices.add(info)
            }
        } catch (e: Exception) {}
    }

    override fun inquiryCompleted(discType: Int) {
        synchronized(lock) { lock.notify() }
    }
    override fun serviceSearchCompleted(transID: Int, respCode: Int) {}
    override fun servicesDiscovered(transID: Int, servRecord: Array<ServiceRecord>) {}

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            var timeout = 10
            var filter: String? = null
            var history: String? = null
            var export: String? = null
            var i = 0
            while (i < args.size) {
                when (args[i]) {
                    "--timeout" -> timeout = args[++i].toInt()
                    "--filter" -> filter = args[++i]
                    "--history" -> history = args[++i]
                    "--export" -> export = args[++i]
                }
                i++
            }
            val scanner = BluetoothScanner(timeout, filter, history, export)
            scanner.startScan()
            // Export CSV if needed
        }
    }
}
